File: LC33_Search in Rotated Sorted Array_Binary Search.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) { // 二分搜索法   往往来说，这种有序数列的查找就需要二分搜索
        int n = nums.size();
        if (n == 0) return -1;
        if (n == 1) return nums[0] == target ? 0:-1;
        int left = 0, right = n - 1, mid;
        while (left <= right) { // 二分搜索
            mid = left + (right - left) / 2; // 这样赋值更好
            if (nums[mid] == target) return mid; // 找到了
            if (nums[0] <= nums[mid]) { // 二分    ！特别注意！：这里要取等号，因为mid可能等于0
                if ((nums[0] <= target) && (target < nums[mid])) right = mid - 1;
                else left =  mid + 1;
            }
            else {
                if ((nums[mid] < target) && (target <= nums[n - 1])) left = mid + 1;
                else right = mid - 1;                
            }
        } 
        return -1;    
    }
};

// reference https://leetcode-cn.com/problems/search-in-rotated-sorted-array/solution/sou-suo-xuan-zhuan-pai-xu-shu-zu-by-leetcode-solut/
// 本题虽然中间某点开始顺序有一定混乱，但分割成两部分，各个部分都是顺序的。所以本题还是可以使用二分法。但使用过程中，要记得把判断条件写好。
// 题解截图见Solution

File: LC33_Search in Rotated Sorted Array_Binary Search.java
class Solution {
    public int search(int[] nums, int target) { // 思路和力扣88相同，但比力扣88简单，因为前置操作少
        int n = nums.length;
        int idx = 0;
        for (int i = 0; i < n - 1; i++) { // 找断点
            if (nums[i] > nums[i + 1]) {
                idx = i;
                break;
            }
        }
        int ans = find(nums, 0, idx, target); // 断点的前半段二分
        if (ans != -1) return ans;
        ans = find(nums, idx + 1, n - 1, target); // 断点的后半段二分
        // 上面的ans前对于idx是否限制主要取决于下面的二分模板怎么写，是return nums[l]还是nums[r]
        return ans;
    }
    int find(int[] nums, int l, int r, int target) {
        while (l < r) {
            int mid = l + r >> 1;
            if (nums[mid] >= target) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }
        return nums[r] == target ? l : -1; // 如果这边是nums[l]，前面第二个ans前
        // 需要加判断，就是控制idx的数值不要越界
    }
}

// reference https://leetcode-cn.com/problems/search-in-rotated-sorted-array/solution/shua-chuan-lc-yan-ge-ologn100yi-qi-kan-q-xifo/